sa/sa_test.go

```diff
@@ -2574,3 +2578,139 @@ func TestIncidentsForSerial(t *testing.T) {
 	test.AssertNotError(t, err, "Failed to retrieve incidents for serial")
 	test.Assert(t, len(incidentsForSerial) == 1, "No active incidents returned")
 }
+
+type inmemSA struct {
+	sapb.StorageAuthorityClient
+	Impl *SQLStorageAuthority
+}
+
+type sfiData struct {
+	serial *sapb.IncidentSerial
+	err    error
+}
+
+type sfiClient struct {
+	// grpc.ClientStream is embedded to meet the
+	// StorageAuthority_SerialsForIncidentClient interface.
+	grpc.ClientStream
+	sfiChan <-chan sfiData
+}
+
+// Recv receives a single serial from the server.
+func (c sfiClient) Recv() (*sapb.IncidentSerial, error) {
+	sfiData := <-c.sfiChan
+	return sfiData.serial, sfiData.err
+}
+
+type sfiServer struct {
+	// grpc.ServerStream is embedded to meet the
+	// StorageAuthority_SerialsForIncidentServer interface.
+	grpc.ServerStream
+	context context.Context
+	sfiChan chan<- sfiData
+}
+
+// Send sends a single serial to the client.
+func (s sfiServer) Send(serial *sapb.IncidentSerial) error {
+	s.sfiChan <- sfiData{serial, nil}
+	return nil
+}
+
+// Context returns the context originally passed by the client.
+func (s sfiServer) Context() context.Context {
+	return s.context
+}
+
+func (s sfiServer) sendErrAndClose(err error) error {
+	s.sfiChan <- sfiData{nil, err}
+	close(s.sfiChan)
+	return nil
+}
+
+func makeinmemSAGRPCStreamAdapter(ctx context.Context) (sfiClient, sfiServer) {
```

Comment:
Instead of the concrete structs `sfiClient` / `sfiServer`, can these return the interface type implemented by those structs? That would serve as implicit documentation of the interface we're going for, and as a nice side effect would enforce that those structs implement the appropriate interfaces.

Reply:
I've made some small changes to allow for this.